lib/jellygrinder/client/helpers/utils.ex

```diff
@@ -0,0 +1,61 @@
+defmodule Jellygrinder.Client.Helpers.Utils do
+  @moduledoc false
+
+  alias Jellygrinder.Client.Helpers.ConnectionManager
+  alias Jellygrinder.Coordinator
+
+  @spec request(String.t(), String.t(), map()) :: {:ok | :error, binary()}
+  def request(path, label, state) do
+    timestamp = get_current_timestamp_ms()
+    start_time = System.monotonic_time()
+    maybe_response = ConnectionManager.get(state.conn_manager, path)
+    end_time = System.monotonic_time()
+
+    request_info = %{
+      timestamp: timestamp,
+      elapsed: System.convert_time_unit(end_time - start_time, :native, :millisecond),
+      label: label,
+      process_name: state.name,
+      path: path
+    }
+
+    {result, data} =
+      case maybe_response do
+        {:ok, response} ->
+          success = response.status == 200
+          data = Map.get(response, :data, "")
+
+          {%{
+             response_code: response.status,
+             success: success,
+             failure_msg: if(success, do: "", else: data),
+             bytes: byte_size(data)
+           }, data}
+
+        {:error, reason} ->
+          {%{
+             response_code: -1,
```

Comment:
Why don't we have a response code? If a request fails, we should receive 400+ code or 500+ code. I would check that.

Reply:
That's the case where we don't get a response at all, so no response code to put here

Reply:
Hmm, but when it is possible because even if we can't reach the host, we should receive the status code [408 in case of timeout](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/408) or some other 500+. So the only possible case I see is when we couldn't even make a request, but in that case, I wouldn't count it in statistics.

Reply:
From the page you linked:
*"It is sent on an idle connection by some servers, even without any previous request by the client."*

What we have here is the following:
* Request couldn't be made
* Request was made but didn't reach the server
* Request was made and reached the server, but server didn't respond. This does not mean we have a timeout, with HTTP/2 we can (and do) also get errors like `%Mint.HTTPError{reason: :unprocessed}`

In all of these cases, server **does not respond at all**. How do you want to infer the status code of a nonexistent response?

These cases absolutely have to be included in the report, as they indicate that e.g. the server can no longer process new requests. Right now, I have this happening consistently whenever I'm exceeding ~410 connections...

Reply:
Ok, I didn't understand it correctly, but in that case, I think that we shouldn't pass `-1` and pass some atom e.g: `:not_responded`. Because this `-1` looks like C-style error handling and in Elixir I believe we could do that better.

Reply:
I agree, but it's for an integer field in a CSV